File: win-prog/ProgWinCS/Dialog Boxes/DialogsWithRegistry/DialogsWithRegistry.cs
//--------------------------------------------------
// DialogsWithRegistry.cs © 2001 by Charles Petzold
//--------------------------------------------------
using Microsoft.Win32;
using System;
using System.Drawing;
using System.Windows.Forms;

class DialogsWithRegistry: BetterFontAndColorDialogs
{
     const string strRegKey = 
                         "Software\\ProgrammingsWindowsWithCSharp\\DialogsWithRegistry";
     const string strFontFace  = "FontFace";
     const string strFontSize  = "FontSize";
     const string strFontStyle = "FontStyle";
     const string strForeColor = "ForeColor";
     const string strBackColor = "BackColor";
     const string strCustomClr = "CustomColor";

     public new static void Main()
     {
          Application.Run(new DialogsWithRegistry());
     }
     public DialogsWithRegistry()
     {
          Text = "Font and Color Dialogs with Registry";

          RegistryKey regkey = Registry.CurrentUser.OpenSubKey(strRegKey);

          if (regkey != null)
          {
               Font = new Font((string) regkey.GetValue(strFontFace),
                               float.Parse(
                                   (string) regkey.GetValue(strFontSize)),
                               (FontStyle) regkey.GetValue(strFontStyle));

               ForeColor = Color.FromArgb(
                              (int) regkey.GetValue(strForeColor));

               BackColor = Color.FromArgb(
                              (int) regkey.GetValue(strBackColor));

               int[] aiColors = new int[16];

               for (int i = 0; i < 16; i++)
                    aiColors[i] = (int) regkey.GetValue(strCustomClr + i);

               clrdlg.CustomColors = aiColors;

               regkey.Close();
          }
     }
     protected override void OnClosed(EventArgs ea)
     {
          RegistryKey regkey = 
                         Registry.CurrentUser.OpenSubKey(strRegKey, true);
          if (regkey == null)
               regkey = Registry.CurrentUser.CreateSubKey(strRegKey);

          regkey.SetValue(strFontFace,  Font.Name);
          regkey.SetValue(strFontSize,  Font.SizeInPoints.ToString());
          regkey.SetValue(strFontStyle, (int) Font.Style);
          regkey.SetValue(strForeColor, ForeColor.ToArgb());
          regkey.SetValue(strBackColor, BackColor.ToArgb());

          for (int i = 0; i < 16; i++)
               regkey.SetValue(strCustomClr + i, clrdlg.CustomColors[i]);

          regkey.Close();
     }
}

File: win-prog/ProgWinVB/Dialog Boxes/DialogsWithRegistry/DialogsWithRegistry.vb
'----------------------------------------------------
' DialogsWithRegistry.vb (c) 2002 by Charles Petzold
'----------------------------------------------------
Imports Microsoft.Win32
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class DialogsWithRegistry
    Inherits BetterFontAndColorDialogs

    Const strRegKey As String = _
            "Software\ProgrammingWindowsWithVBdotNet\DialogsWithRegistry"
    Const strFontFace As String = "FontFace"
    Const strFontSize As String = "FontSize"
    Const strFontStyle As String = "FontStyle"
    Const strForeColor As String = "ForeColor"
    Const strBackColor As String = "BackColor"
    Const strCustomClr As String = "CustomColor"

    Shared Shadows Sub Main()
        Application.Run(New DialogsWithRegistry())
    End Sub

    Sub New()
        Text = "Font and Color Dialogs with Registry"

        Dim regkey As RegistryKey = _
                            Registry.CurrentUser.OpenSubKey(strRegKey)

        If Not regkey Is Nothing Then
            Font = New Font( _
                    DirectCast(regkey.GetValue(strFontFace), String), _
                    Single.Parse( _
                        DirectCast(regkey.GetValue(strFontSize), String)), _
                    CType(regkey.GetValue(strFontStyle), FontStyle))
            ForeColor = Color.FromArgb(CInt(regkey.GetValue(strForeColor)))
            BackColor = Color.FromArgb(CInt(regkey.GetValue(strBackColor)))

            Dim i, aiColors(16) As Integer

            For i = 0 To 15
                aiColors(i) = CInt(regkey.GetValue(strCustomClr & i))
            Next i
            clrdlg.CustomColors = aiColors
            regkey.Close()
        End If
    End Sub

    Protected Overrides Sub OnClosed(ByVal ea As EventArgs)
        Dim i As Integer
        Dim regkey As RegistryKey = _
                    Registry.CurrentUser.OpenSubKey(strRegKey, True)

        If regkey Is Nothing Then
            regkey = Registry.CurrentUser.CreateSubKey(strRegKey)
        End If

        regkey.SetValue(strFontFace, Font.Name)
        regkey.SetValue(strFontSize, Font.SizeInPoints.ToString())
        regkey.SetValue(strFontStyle, CInt(Font.Style))
        regkey.SetValue(strForeColor, ForeColor.ToArgb())
        regkey.SetValue(strBackColor, BackColor.ToArgb())

        For i = 0 To 15
            regkey.SetValue(strCustomClr & i, clrdlg.CustomColors(i))
        Next i
        regkey.Close()
    End Sub
End Class
